### PR TITLE
[Bugfix][Core] Output sampling: heuristic to choose between candidates

### PR DESCRIPTION
**Pull request by NihalPotdar**
In the `RequestOutput` class in `outputs.py`, when beam search is not being used, VLLM samples from the best-of generated candidates with the `cumulative_logprob` heuristic. Given that log_probs are negative and then we sample from these candidates from highest (closest to zero) to smallest (farthest from zero), the chosen solutions would bias towards shorter sequences. A better heuristic would likely be to average out the cumulative log_probs by the sequence length as suggested in this PR.

Validation: running some quality-based benchmarks would be helpful in assessing the effectiveness of this change. Are there any which are already built into VLLM?

### CODE REVIEW

**Comment by NihalPotdar** [comment]
cc @zhuohan123 can you please take a look at this? ty

**Comment by zhuohan123** [comment]
I would suggest keep the original semantics. The task of language modeling is to model the probability of a whole sentence. Like:

```
P(A B <eos>) = P(A)P(B | A)P(<eos> | A B) = exp(log(P(A)) + log(P(B | A)) + log(P(<eos> | A B)))
P(A <eos>) = P(A)P(<eos> | A) = y = exp(log(P(A)) + log(P(<eos> | A)))
...
```
And then if `log(P(A)) + log(P(B | A)) + log(P(<eos> | A B))` is larger than `log(P(A)) + log(P(<eos> | A))`, we should pick the first sentence because it has higher probability according to the language model.

If you change the semantics to average, you are not respecting the original distribution. Instead, you are selecting the best one based on
```
(log(P(A)) + log(P(B | A)) + log(P(<eos> | A B)))/3 = exp(log(P(A B <eos>)) / 3)
(log(P(A)) + log(P(<eos> | A)))/2 = exp(log(P(A <eos>)) / 2)
```
instead of the original probability.

In short, the summation here is not a heuristic, it's following the LLM's output probability distribution. Changing it to average is actually a heuristic that breaks the original probability distribution.

For your use case, I would suggest you sample n samples, and then perform your own custom `best_of` logic outside of vLLM.

**Comment by NihalPotdar** [comment]
@zhuohan123 this makes sense. In this case, shouldn't the implementation of beam-search and consequently, beam scoring be aligned with this? `get_beam_search_score` in the sequence.py.

**Comment by zhuohan123** [comment]
> @zhuohan123 this makes sense. In this case, shouldn't the implementation of beam-search and consequently, beam scoring be aligned with this? [`get_beam_search_score` in the sequence.py.](https://github.com/vllm-project/vllm/blob/d7f4178dd99cdba8735d268ff470929de53ee2d5/vllm/sequence.py#L357)

Most beam search implementation includes length penalty, which is a technique used to control the length of beam search results: https://d2l.ai/chapter_recurrent-modern/beam-search.html#equation-eq-beam-search-score

**Comment by NihalPotdar** [comment]
Yup, I was reading through the paper which seemingly introduced this implementation of beam search: https://arxiv.org/pdf/1609.08144. In general, what I meant was that their motivation for introducing the length normalization term (1/L) was very similar to the earlier intuition I was describing, not wanting to bias towards shorter sequences. 

> Without some form of length-normalization regular beam search will favor shorter results over longer ones on average since a negative log-probability is added at each step, yielding lower (more negative) scores for longer sentences. We first tried to simply divide by the length to normalize. We then improved on that original heuristic by dividing by length α...

Since the goal of both beam search and parallel sampling is the same - to choose the most optimal sequences among some candidates, wouldn't it make more sense to align how sequences are chosen between these approaches? or perhaps offer a callable function as a parameter to the user to configure their custom scoring function for parallel sampling?

Just looking through what some other OSS libraries do, it seems that a lot only support either beam search and when they support parallel sampling (best-of-N sampling) they offer a function for the user to specify their scoring function:
- hugging face transformers does not support this parallel sampling, but it is supported by TRL [which exposes a scoring function to the user](url)
- trt-llm does not support best-of-N sampling, [only beam search or generate one candidate](https://huggingface.co/docs/transformers/v4.43.2/en/generation_strategies)

btw thank you all for the replies!

**Comment by NihalPotdar** [comment]
hi @zhuohan123 just bumping this again -- curious if you have any further thoughts on this

**Comment by kzhang-dm** [comment]
@NihalPotdar Agree with you. I recently have the same concern.   When using beam search, the top n candidates should be selected based on ```length normalized lobprob``` but not ```cumulative_logprob ```. The later give undesired performance for using beam search.    One suggest would be adding an optional argument to ```SamplingParams```, so user can optionally rank the candidates with length normalized or unnormalized score.    

cc @zhuohan123

**Comment by kzhang-dm** [comment]
> @zhuohan123 this makes sense. In this case, shouldn't the implementation of beam-search and consequently, beam scoring be aligned with this? [`get_beam_search_score` in the sequence.py.](https://github.com/vllm-project/vllm/blob/d7f4178dd99cdba8735d268ff470929de53ee2d5/vllm/sequence.py#L357)

Also realize that the `get_beam_search_score()` use the sum of `len(self._output_token_ids) + len(self._prompt_token_ids)` to normalize the `cumulative_logprob', this seems to be a bug: 
https://github.com/vllm-project/vllm/blob/d7f4178dd99cdba8735d268ff470929de53ee2d5/vllm/sequence.py#L162
